samza-test/src/main/java/org/apache/samza/test/framework/MessageStreamAssert.java

```diff
@@ -0,0 +1,192 @@
+  private final Serde<M> serde;
+  private boolean checkEachTask = false;
+
+  /**
+   * Constructors a MessageStreamAssert with an id and serde
+   * @param id unique id
+   * @param messageStream represents messageStream that you want to assert on
+   * @param serde serde used to desialize messageStream
+   * @param <M> represents type of Message
+   * @return MessageStreamAssert that returns the the messages in the stream
+   */
+  public static <M> MessageStreamAssert<M> that(String id, MessageStream<M> messageStream, Serde<M> serde) {
+    return new MessageStreamAssert<>(id, messageStream, serde);
+  }
+
+  private MessageStreamAssert(String id, MessageStream<M> messageStream, Serde<M> serde) {
+    this.id = id;
+    this.messageStream = messageStream;
+    this.serde = serde;
+  }
+
+  public MessageStreamAssert forEachTask() {
+    checkEachTask = true;
+    return this;
+  }
+
+  public void containsInAnyOrder(final Collection<M> expected) {
+    LATCHES.putIfAbsent(id, PLACE_HOLDER);
+    final MessageStream<M> streamToCheck = checkEachTask
+        ? messageStream
+        : messageStream
+            .partitionBy(m -> null, m -> m, KVSerde.of(new StringSerde(), serde), null)
+            .map(kv -> kv.value);
+
+    streamToCheck.sink(new CheckAgainstExpected<M>(id, expected, checkEachTask));
+  }
+
+  public static void waitForComplete() {
+    try {
+      while (!LATCHES.isEmpty()) {
+        final Set<String> ids  = new HashSet<>(LATCHES.keySet());
+        for (String id : ids) {
+          while (LATCHES.get(id) == PLACE_HOLDER) {
+            Thread.sleep(100);
+          }
+
+          final CountDownLatch latch = LATCHES.get(id);
+          if (latch != null) {
+            latch.await();
+            LATCHES.remove(id);
+          }
+        }
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static final class CheckAgainstExpected<M> implements SinkFunction<M> {
+    private static final long TIMEOUT = 5000L;
+
+    private final String id;
+    private final boolean checkEachTask;
+    private final transient Collection<M> expected;
+
+
+    private transient Timer timer = new Timer();
+    private transient List<M> actual = Collections.synchronizedList(new ArrayList<>());
+    private transient TimerTask timerTask = new TimerTask() {
+      @Override
+      public void run() {
+        check();
+      }
+    };
+
+    CheckAgainstExpected(String id, Collection<M> expected, boolean checkEachTask) {
+      this.id = id;
+      this.expected = expected;
+      this.checkEachTask = checkEachTask;
+    }
+
+    @Override
+    public void init(Config config, TaskContext context) {
+      final SystemStreamPartition ssp = Iterables.getFirst(context.getSystemStreamPartitions(), null);
+      if (ssp != null || ssp.getPartition().getPartitionId() == 0) {
+        final int count = checkEachTask ? context.getSamzaContainerContext().taskNames.size() : 1;
```

Comment:
this condition doesn't make sense!

Reply:
https://issues.apache.org/jira/browse/SAMZA-1798